### DISCUSSION TITLE
declared_attr typing

### DISCUSSION THREAD

**Discussion by prousso** [sqlalchemy/sqlalchemy#12193] [Usage Questions] [answered]
Hi, I have some issues with declared_attr and mypy.
```python
from sqlalchemy.orm import Mapped, declared_attr, relationship
from sqlalchemy.orm.decl_api import (
    DeclarativeBase,
)

class SqlAlchemyBase(DeclarativeBase):
    pass


class Flower(SqlAlchemyBase):
    id: Mapped[int]

class BigCactus(SqlAlchemyBase):
    id: Mapped[int]

    @declared_attr
    def group(self) -> Mapped[Flower]:
        return relationship(Flower)

class LittleCactus(BigCactus):
    group = relationship(Flower)
```

Which result with
```bash
cactus.py:21:13: error: Incompatible types in assignment (expression has type "_RelationshipDeclared[Any]", base class "BigCactus" defined the type as
"Callable[[BigCactus], Mapped[Flower]]")  [assignment]
```
I also tried Mapped[RelationshipProperty] but it did not solve my issue.
What is my mistake here?

Thanks

**Comment by zzzeek** [comment, marked as answer]
I wouldnt recommend overriding a `@declared_attr` with a plain attribute like that, I dont think we have any tests for that format.

beyond that, Python typing is really not going to let you override a Python descriptor with a plain attribute like that.   typing doesnt really handle this case and there is nothing SQLAlchemy can do about taht

in the example given there's no need for BigCactus.group to be a declared_attr in the first place, but I would probably maintain the use of `@declared_attr` for a subclass that wants to override it also if in fact the base class needs to use `@declared_attr` for some reason

alternatively you can just use TYPE_CHECKING or `ignore` to override

```py
class LittleCactus(BigCactus):
    if not TYPE_CHECKING:
        group: Mapped[Flower] = relationship(Flower)
```


here's a demo of how this is not possible with Python typing

```py
from typing import TypeVar, Callable, Any, Generic

_C = TypeVar("_C", bound=Callable[[Any], Any])
_T = TypeVar("_T", bound=Any)


class mydescriptor(Generic[_T]):
    def __init__(self, fn: Callable[..., _T]):
        self.fn = fn

    def __get__(self, obj: Any, type: Any = None) -> _T:
        return self.fn(obj)


def mydecorator(fn: _C) -> mydescriptor[_C]:
    return mydescriptor(fn)


class Foo:
    @mydecorator
    def bar(self) -> str:
        return "hi"


class Bar(Foo):
    bar: str = "hello"

```

same error:

```
test4.py:23: error: Incompatible types in assignment (expression has type "str", base class "Foo" defined the type as "Callable[[Foo], str]")  [assignment]

```
